### PR TITLE
Docs for size_of::<#[repr(C)]> items.

### PR DESCRIPTION
**Pull request by Havvy**
Most of this info comes from camlorn's blog post on optimizing struct layout and the Rustonomicon.

I don't really like my wording in the first paragraph.

I also cannot find a definition of what `#[repr(C)]` does for enums that have variants with fields. They're allowed, unlike `#[repr(C)] enum`s with no variants.

### CODE REVIEW

**Comment by Havvy** [comment]
r? @steveklabnik 

(Why do I always forget this?)

File: src/libcore/mem.rs [hunk 1]
```diff
@@ -209,6 +209,35 @@ pub fn forget<T>(t: T) {
 /// The mutability of a pointer does not change its size. As such, `&T` and `&mut T`
 /// have the same size. Likewise for `*const T` and `*mut T`.
 ///
+/// # Size of #[repr(C)] items
```
Comment (kennytm) [inline src/libcore/mem.rs]: Put `#[repr(C)]` inside backticks `` `…` ``.

File: src/libcore/mem.rs [hunk 2]
```diff
@@ -209,6 +209,35 @@ pub fn forget<T>(t: T) {
...
+/// Finally, round the size of the struct to the nearest multiple of its [alignment].
+///
+/// Unlike `C`, zero sized structs are not rounded up to one byte in size.
+///
+/// ## Enums
+///
+/// Enums that carry no data other than the descriminant have the same size as C enums
+/// on the platform they are compiled for.
+///
+/// ## Unions
```
Comment (kennytm) [inline src/libcore/mem.rs]: I don't know what's wrong with linkchecker, but it seems this title is not acceptable.

```
[01:09:19] std/mem/index.html:32: broken link fragment `#unions` pointing to `std/mem/index.html`
[01:09:25] core/mem/index.html:32: broken link fragment `#unions` pointing to `core/mem/index.html`
[01:09:28] thread 'main' panicked at 'found some broken links', /checkout/src/tools/linkchecker/main.rs:49:8
```

Comment (kennytm) [reply src/libcore/mem.rs]: @Havvy Ok the problem is that `ManuallyDrop` is a union, so there's already a "Unions" section in the documentation (check https://doc.rust-lang.org/std/mem/index.html). The linkchecker error message should say something like "duplicated link ID". 

Could we workaround by saying "Size of Structs"/"Size of Enums"/"Size of Unions"?

**Comment by steveklabnik** [comment]
@bors: r+ rollup

thanks! ❤️

**Pull request merged**